Implement parallel collect for HashMap

edre opened the pull request:
test rayon_collect_parallel ... bench:  28,819,674 ns/iter (+/- 1,225,005)
    test rayon_collect_serial   ... bench:  57,375,202 ns/iter (+/- 1,833,072)
    test rayon_set_sum_parallel ... bench:   1,202,368 ns/iter (+/- 70,364)
    test rayon_set_sum_serial   ... bench:   7,130,543 ns/iter (+/- 165,480)
    test std_collect_parallel   ... bench:  75,311,224 ns/iter (+/- 22,114,519)
    test std_collect_serial     ... bench:  61,104,321 ns/iter (+/- 5,774,424)
    test std_set_sum_parallel   ... bench:  10,787,688 ns/iter (+/- 668,539)
    test std_set_sum_serial     ... bench:   9,239,331 ns/iter (+/- 284,888)

Surprisingly, serial collect and serial set_sum both become notably faster, presumably just from the indexing shift in the first commit. No idea why that would give different results.

edre committed:
Derive table address from hash using upper bits instead of lower bits.

This allows us to preshard to a fixed number of shards without knowing ahead
of time how large the table will be, as the prefix of the table address will
always be contiguous.

Fixed resize to do the normal Robin Hood insertion, as now the order can
change.

edre committed:
Implement parallel collect for HashMap.

Works over 3 phases:
1) Hash and partition items into shard buckets.
2) Insert shard buckets into their segment of the hash table in parallel.
3) Fix up any entries that needed to overflow into the next segment.

test rayon_collect_parallel ... bench:  28,819,674 ns/iter (+/- 1,225,005)
test rayon_collect_serial   ... bench:  57,375,202 ns/iter (+/- 1,833,072)
test std_collect_parallel   ... bench:  75,311,224 ns/iter (+/- 22,114,519)
test std_collect_serial     ... bench:  61,104,321 ns/iter (+/- 5,774,424)

Fixes #17

Future work: implement for HashSet, modify to work with ParallelExtend

edre commented:
Serial collect is probably faster because resizing is now more cache local. The two places a value could be moved to in the bigger table are now adjacent instead of far away.

edre commented:
The test failures (that do not occur on _my_ machine) seem to suggest that num_cpus::get() and SafeHash are of different bit sizes on the test target, but they both look like usize to me.

cuviper commented on src/par/map.rs:
```diff
@@ -74,13 +78,95 @@ impl<'a, K: Sync, V: Send, S> IntoParallelIterator for &'a mut HashMap<K, V, S>
+        let mut map = HashMap::<K, V, S>::default();
+        // One shard per core, but round up to a power of two to keep the math simpler.
+        let num_shards = (num_cpus::get() - 1).next_power_of_two();
+        let shard_shift = num_shards.leading_zeros() + 1;
```
If `num_cpus::get() == 1`, then `shard_shift` will equal the number of bits in `usize`, which will overflow in `Shr`.

cuviper replied:
Also, `next_power_of_two()` returns a value greater *or equal* -- e.g. 0 -> 1, 1 -> 1, 2 -> 2, 3 -> 4, ... so I don't think you really want to subtract 1 first.

edre committed:
Fix collect sharding for num_cpus=1

edre committed:
Add FNV hash comparison for collect benchmark.

Our hash addressing scheme takes the higher bits, which most fast hash
functions don't prioritize in their mixing. The results show that indeed
std_hash is again faster with FNV, due to collision avoidance.

test rayon_collect_fnv      ... bench:  94,820,409 ns/iter (+/- 681,480)
test rayon_collect_serial   ... bench: 110,255,836 ns/iter (+/- 5,174,574)
test std_collect_fnv        ... bench:  87,217,501 ns/iter (+/- 419,949)
test std_collect_serial     ... bench: 112,992,257 ns/iter (+/- 1,418,915)

edre committed:
Fix parallelization bottlenecks for collect.

1) Don't write to the table size field in the inner loop of every thread.
2) Move the table initializion to each segment.
3) Move the element counting to be in parallel.

test rayon_collect_fnv      ... bench:  49,981,752 ns/iter (+/- 1,238,650)
test rayon_collect_parallel ... bench:  14,401,416 ns/iter (+/- 2,051,357)
test rayon_collect_serial   ... bench:  61,508,596 ns/iter (+/- 1,182,697)
test std_collect_fnv        ... bench:  51,129,763 ns/iter (+/- 2,093,351)
test std_collect_parallel   ... bench:  76,346,500 ns/iter (+/- 15,314,104)
test std_collect_serial     ... bench:  62,928,543 ns/iter (+/- 1,286,422)

edre commented:
I added a few tweaks to resolve why I wasn't getting more than ~2x speedup, and now it's most of the way to linearly parallelizable. There are a bit more hooks into the internal system than I initially expected, but they're most in the form of extra types methods, that should merge easily to upstream changes.

cuviper commented:
Sorry we never landed this, but I'm deprecating this crate now that [`hashbrown`](https://github.com/Amanieu/hashbrown) has replaced the standard implementation. They do have `rayon` impls in that crate too, with a similarly unoptimized `ParallelExtend`, so there may be room for improvement there. I'm sure your design would need to be reworked for the new table format though, if the idea can work at all...

edre commented:
Looking through the hashbrown implementation, and the same basic idea should work, but quadratic probing may make it less effective in practice. More values are likely to overflow to the next parallel shard.
Exciting that the rayon extension is being done in the main crate.

cuviper commented:
> Exciting that the rayon extension is being done in the main crate.

Yes, but note that it doesn't mean the `std` types will have that feature -- you'll still have to use the external `hashbrown` crate for proper rayon support. We need `std` in rayon, so we can't also be a dependency therein.

It would be neat if we could expose some simple API in `std` to make it possible for `rayon` to implement the improved parallel iterators for those types.
